### PR TITLE
[4.0][API][com_actionslogs] keep track of API requests

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes
listen event `onAfterDispatch` to keep track of API requests on actionlogs 

### Testing Instructions
make some requests GET for example:
/api/index.php/v1/plugins
/api/index.php/v1/users
/api/index.php/v1/config/application
/api/index.php/v1/content/article
....
etc...

test other verbs as welll

### Expected result
API requests logged 
something like 
Screenshot from 2019-11-17 11-12-02

### Actual result

N/A

### Documentation Changes Required
maybe

### CODE REVIEW

File: plugins/actionlog/joomla/joomla.php [hunk 1]
```diff
@@ -1041,4 +1046,33 @@ public function onAfterPurge($group = 'all')
...
+		$user    = Factory::getUser();
+
+		if (!$this->checkLoggable($context))
+		{
+			return;
+		}
+
+		$message = array(
+			'action'      => 'API',
+			'verb'        => $_SERVER['REQUEST_METHOD'],
```
Comment (mbabker) [inline plugins/actionlog/joomla/joomla.php]: Get the request method from the Input object.

File: plugins/actionlog/joomla/joomla.php [hunk 2]
```diff
@@ -1041,4 +1046,33 @@ public function onAfterPurge($group = 'all')
...
+			'username'    => $user->username,
+			'accountlink' => 'index.php?option=com_users&task=user.edit&id=' . $user->id,
+			'url'         => $_SERVER['PATH_INFO'],
```
Comment (mbabker) [inline plugins/actionlog/joomla/joomla.php]: If the API app runs the same as the regular web app, the route should be available there.  Check something like `$this->app->get('url')`.

Basically, don’t direct use the superglobals.

**Comment by mbabker** [comment]
Does this REALLY need a new event?  Couldn’t `onAfterDispatch` work just as well here?  You’d just need to check context if using that event.

File: plugins/actionlog/joomla/joomla.php (hunk 1 above)
Comment (alikon) [reply plugins/actionlog/joomla/joomla.php]: yes, better

File: plugins/actionlog/joomla/joomla.php (hunk 2 above)
Comment (alikon) [reply plugins/actionlog/joomla/joomla.php]: right

**Comment by alikon** [comment]
yes again :smile: 
there is no need for a new event `onAfterDispatch` does already the job

File: plugins/actionlog/joomla/joomla.php [hunk 3]
```diff
@@ -1041,4 +1046,38 @@ public function onAfterPurge($group = 'all')
...
+			return;
+		}
+
+		$user = $this->app->getIdentity();
+		$message = array(
+			'action'      => 'API',
+			'verb'        => $this->app->input->server->getString('REQUEST_METHOD', ''),
+			'username'    => $user->username,
+			'accountlink' => 'index.php?option=com_users&task=user.edit&id=' . $user->id,
+			'url'         => $this->app->input->server->getString('PATH_INFO', ''),
```
Comment (wilsonge) [inline plugins/actionlog/joomla/joomla.php]: This won't work in IIS. Use the value from the app (see https://github.com/joomla-framework/application/blob/2.0-dev/src/AbstractWebApplication.php#L997)

`$this->app->get('uri.base.path')`

File: plugins/actionlog/joomla/joomla.php [hunk 4]
```diff
@@ -1041,4 +1046,38 @@ public function onAfterPurge($group = 'all')
...
+
+		if (!$this->checkLoggable($context))
+		{
+			return;
+		}
+
+		$user = $this->app->getIdentity();
+		$message = array(
+			'action'      => 'API',
+			'verb'        => $this->app->input->server->getString('REQUEST_METHOD', ''),
```
Comment (wilsonge) [inline plugins/actionlog/joomla/joomla.php]: `$this->app->input->getMethod()` (https://github.com/joomla-framework/input/blob/2.0-dev/src/Input.php#L350)

File: plugins/actionlog/joomla/joomla.php (hunk 3 above)
Comment (alikon) [reply plugins/actionlog/joomla/joomla.php]: done

File: plugins/actionlog/joomla/joomla.php (hunk 4 above)
Comment (alikon) [reply plugins/actionlog/joomla/joomla.php]: done

**Comment by wilsonge** [comment]
Looks good to me. Just needs some tests

File: plugins/actionlog/joomla/joomla.php [hunk 5]
```diff
@@ -766,6 +766,11 @@ public function onUserAfterDeleteGroup($group, $success, $msg)
 	 */
 	public function onUserAfterLogin($options)
 	{
+		if ($options['action'] = 'core.api.login')
```
Comment (mbabker) [inline plugins/actionlog/joomla/joomla.php]: You’re assigning a value here.  Surely you meant to do a comparison?

Comment (alikon) [reply plugins/actionlog/joomla/joomla.php]: two error in one row 
:sob:

**Comment by Quy** [comment]
Please fix conflicts.

**Comment by alikon** [comment]
conflict solved

**Comment by Quy** [comment]
> User admin performed a GET on index.php/v1/modules/site?language=%2A&state=0

`*` is displayed as `%2A`

File: plugins/actionlog/joomla/joomla.php [hunk 6]
```diff
@@ -1076,7 +1076,7 @@ public function onAfterDispatch()
 			'verb'        => $this->app->input->getMethod(),
 			'username'    => $user->username,
 			'accountlink' => 'index.php?option=com_users&task=user.edit&id=' . $user->id,
-			'url'         => $this->app->get('uri.route'),
+			'url'         => urldecode($this->app->get('uri.route')),
```
Comment (Quy) [inline plugins/actionlog/joomla/joomla.php]: This can output potential XSS.

Comment (alikon) [reply plugins/actionlog/joomla/joomla.php]: any suggestions ?
what about 
```php
urldecode(htmlspecialchars($this->app->get('uri.route'), ENT_QUOTES, 'UTF-8'));
```

**Comment by chmst** [comment]
Tested with different words, for example 


api-test-actionlogs

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on afb8f8fdc20b87dca3ea0ff3ff0a93ef032c1e2e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27094">issues.joomla.org/tracker/joomla-cms/27094</a>.</sub>

**Comment by wilsonge** [comment]
Having thought about this I think we probably want to have an additional flag for this API setting. Because in the case you have mobile apps using the API actively you're going to end up having too many entries in action logs. But you might well still want other actions logged in the component.

**Comment by N6REJ** [comment]
Tests ran:
image

Results:
image


I have tested this item :white_check_mark: successfully on afb8f8fdc20b87dca3ea0ff3ff0a93ef032c1e2e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27094">issues.joomla.org/tracker/joomla-cms/27094</a>.</sub>

**Comment by alikon** [comment]
ok i've fine grained the api logging
Screenshot from 2020-03-21 09-35-45

now you can disable the api logging
you can enable only some verbs to be logged
and of course only on some "Events"

**Comment by alikon** [comment]
@chmst, @N6REJ  another round of tests please

File: administrator/language/en-GB/com_actionlogs.ini [hunk 7]
```diff
@@ -9,6 +9,7 @@ COM_ACTIONLOGS_ACTION_ASC="Action ascending"
 COM_ACTIONLOGS_ACTION_DESC="Action descending"
 COM_ACTIONLOGS_ACTION_VIEWLOGS="View logs"
 COM_ACTIONLOGS_ACTION_VIEWLOGS_DESC="Allows users in the group to view the logs."
+COM_ACTIONLOGS_API_LOGGING_LABEL="Log API requests"
```
Comment (Quy) [inline administrator/language/en-GB/com_actionlogs.ini]:
```suggestion
COM_ACTIONLOGS_API_LOGGING_LABEL="Log API Requests"
```

File: administrator/language/en-GB/com_actionlogs.ini [hunk 8]
```diff
@@ -22,13 +23,19 @@ COM_ACTIONLOGS_EXTENSION="Extension"
...
+COM_ACTIONLOGS_FIELD_VALUE_POST="POST"
+COM_ACTIONLOGS_FIELD_VALUE_PUT="PUT"
 COM_ACTIONLOGS_FILTER_SEARCH_DESC="Search in username. Prefix with ID: to search for an actionlog ID. Prefix with ITEM_ID: to search for an actionlog item ID."
 COM_ACTIONLOGS_IP_ADDRESS="IP Address"
 COM_ACTIONLOGS_IP_ADDRESS_ASC="IP Address ascending"
 COM_ACTIONLOGS_IP_ADDRESS_DESC="IP Address descending"
 COM_ACTIONLOGS_IP_INVALID="Invalid IP"
 COM_ACTIONLOGS_IP_LOGGING_LABEL="IP Logging"
 COM_ACTIONLOGS_LOG_EXTENSIONS_LABEL="Events To Log"
+COM_ACTIONLOGS_LOG_VERBS_LABEL="Verbs to Log"
```
Comment (Quy) [inline administrator/language/en-GB/com_actionlogs.ini]:
```suggestion
COM_ACTIONLOGS_LOG_VERBS_LABEL="Verbs To Log"
```

File: plugins/actionlog/joomla/joomla.php (hunk 6 above)
Comment (Quy) [inline plugins/actionlog/joomla/joomla.php]:
```suggestion
			'url'         => htmlspecialchars(urldecode($this->app->get('uri.route')), ENT_QUOTES, 'UTF-8'),
```

Comment (Quy) [reply plugins/actionlog/joomla/joomla.php]: Here is a test `/api/index.php/v1/plugins?filter[search]=<b>Test</b>`. Make sure `Test` is not bold in the output.

File: administrator/components/com_actionlogs/config.xml [hunk 9]
```diff
@@ -28,5 +28,30 @@
...
+			<option value="0">JNO</option>
+			<option value="1">JYES</option>
+		</field>
+		<field
+			name="loggable_verbs"
+			type="list"
+			label="COM_ACTIONLOGS_LOG_VERBS_LABEL"
+			multiple="true"
+			showon="loggable_api:1"
+			default="GET"
```
Comment (Quy) [inline administrator/components/com_actionlogs/config.xml:48]:
```suggestion
			default="GET"
			validate="options"
```

**Comment by Razzo1987** [comment]
I have tested this item :white_check_mark: successfully on 4a380778bff696a4b703c33d9b575790f8a70885<br /><br />GET: ok
POST: ok
DELETE: ok
PUT: not tested (if someone have an example I try it)
PATCH: non listed in "Verds to log" ( I test this example: https://docs.joomla.org/J4.x:Joomla_Core_APIs#Update_Article)

J API<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27094">issues.joomla.org/tracker/joomla-cms/27094</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Comment by alikon** [comment]
thanks folks + thanks to covid19 for lockdown 
:smile:

**Pull request merged**